### PR TITLE
Desktop: Resolves #9927: Beta editor: Fix search results not highlighted

### LINKED ISSUE #9927: Desktop: Joplin search terms not highlighted in beta CM6 editor or viewer
### Operating system

Linux

### Joplin version

2.14.12

### Desktop version info

_No response_

### Current behaviour

1. Enable the beta editor
2. Search for a term that should have results

**Note**: Currently, search highlighting is only implemented for the CM5 editor.

### Expected behaviour

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by personalizedrefrigerator**
# Summary

- Updates `CodeMirror5Emulation` such that Joplin's CodeMirror 5 search highlighting works with CodeMirror 6.
- For now, search results aren't marked on the scrollbar &mdash; this might be something we need to implement ourselves ([see this post on the CM6 forum](https://discuss.codemirror.net/t/show-location-of-matched-search-query-in-scrollbar/6463)).

Fixes #9927.

# Testing

1. Enable the CodeMirror 5 editor
2. Search for a full-word term known to be present in multiple notes (e.g. "references")
3. Change the search term to work around #9926
4. Verify that matches are shown on the scrollbar
5. Verify that matches are highlighted in both the editor and viewer

6. Enable the CodeMirror 6 editor
7. Search for "references"
8. Change the search term to work around #9926
9. Verify that matches are highlighted in both the editor and viewer

This has been tested successfully on Ubuntu 23.10.

### CODE REVIEW

File: packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useEditorSearchExtension.ts [hunk 1]
```diff
@@ -172,7 +176,7 @@ export default function useEditorSearch(CodeMirror: any) {
 		// These operations are pretty slow, so we won't add use them until the user
 		// has finished typing, 500ms is probably enough time
 		const timeout = shim.setTimeout(() => {
-			const scrollMarks = this.showMatchesOnScrollbar(searchTerm, true, 'cm-search-marker-scrollbar');
+			const scrollMarks = this.showMatchesOnScrollbar?.(searchTerm, true, 'cm-search-marker-scrollbar');
```
Comment (personalizedrefrigerator) [inline packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useEditorSearchExtension.ts:179]: At present, `showMatchesOnScrollbar` is only implemented for CodeMirror 5. As per [this discussion thread](https://discuss.codemirror.net/t/show-location-of-matched-search-query-in-scrollbar/6463), we'll likely need to implement this ourselves.

File: packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useEditorSearchHandler.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	setLocalSearchResultCount(count: number): void;
+	searchMarkers: any;
+	webviewRef: RefObject<any>;
+	editorRef: RefObject<any>;
+
+	noteContent: string;
+	renderedBody: RenderedBody;
+}
+
+const useEditorSearchHandler = (props: Props) => {
```
Comment (personalizedrefrigerator) [inline packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useEditorSearchHandler.ts:16]: `useEditorSearchHandler` was originally part of `v5/CodeMirror.tsx`.

File: packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useEditorSearchExtension.ts [hunk 3]
```diff
@@ -73,15 +77,15 @@ export default function useEditorSearch(CodeMirror: any) {
...
+			match = { from: cursor.from(), to: cursor.to() };
 		}
 
 		if (match) {
 			if (scrollTo) {
 				if (withSelection) {
 					cm.setSelection(match.from, match.to);
 				} else {
-					cm.scrollTo(match);
+					cm.scrollIntoView(match);
```
Comment (personalizedrefrigerator) [inline packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useEditorSearchExtension.ts:88]: According to the CodeMirror documentation, [`.scrollTo` takes two arguments](https://codemirror.net/5/doc/manual.html#scrollTo), while [`scrollIntoView` can take an object with the same type as `match`](https://codemirror.net/5/doc/manual.html#scrollIntoView).

As such, I suspect we were calling `scrollTo` incorrectly.

**Comment by laurent22** [comment]
@personalizedrefrigerator there's one conflict here as well

**Comment by laurent22** [comment]
Sorry, a few more conflicts now

**Comment by laurent22** [comment]
It's complaining about a spelling mistake:

> 2024-03-02T19:22:58.9062623Z /home/runner/work/joplin/joplin/packages/editor/CodeMirror/CodeMirror5Emulation/CodeMirror5Emulation.test.ts:135:58 - Unknown word (Testst)

**Pull request merged**